src/win32crypto/engine.c

```diff
@@ -0,0 +1,605 @@
+    do {
+        DWORD check = CERT_STORE_SIGNATURE_FLAG;
+        PCCERT_CONTEXT local_iss = CertGetIssuerCertificateFromStore(engine->ca, peer_cert, NULL, &check);
+        if (local_iss) {
+            CertFreeCertificateContext(peer_cert);
+            CertFreeCertificateContext(local_iss);
+            return 0;
+        }
+
+        check = 0;
+        peer_cert = CertGetIssuerCertificateFromStore(peer_cert->hCertStore, peer_cert, peer_cert, &check);
+    } while (peer_cert);
+
+    return -1;
+}
+
+static SECURITY_STATUS verify_server_cert(struct win32crypto_engine_s *engine)
+{
+    PCCERT_CONTEXT server_cert = NULL;
+    SECURITY_STATUS rc;
+
+    rc = QueryContextAttributes(&engine->ctxt_handle, SECPKG_ATTR_REMOTE_CERT_CHAIN, &server_cert);
+    if (rc != SEC_E_OK) {
+        LOG_LAST_ERROR(ERR, "failed to get server cert");
+        return rc;
+    }
+
+    tlsuv_certificate_t crt = (tlsuv_certificate_t)win32_new_cert(server_cert, server_cert->hCertStore);
+    int verified = engine->cert_verify_f(crt, engine->verify_ctx);
+    crt->free(crt);
+    return  verified == 0 ? ERROR_SUCCESS : TRUST_E_FAIL;
+}
+
+static tls_handshake_state engine_handshake(tlsuv_engine_t self) {
+    struct win32crypto_engine_s *engine = (struct win32crypto_engine_s *)self;
+
+    if (engine->handshake_st == TLS_HS_COMPLETE ||
+        engine->handshake_st == TLS_HS_ERROR)
+        return engine->handshake_st;
+
+    u_long req_flags =
+            ISC_REQ_USE_SUPPLIED_CREDS |
+            ISC_REQ_ALLOCATE_MEMORY |
+            ISC_REQ_CONFIDENTIALITY |
+            ISC_REQ_REPLAY_DETECT |
+            ISC_REQ_SEQUENCE_DETECT |
+            ISC_REQ_STREAM;
+    u_long ret_flags = 0;
+    SecBuffer outbuf[3] = {
+        { .BufferType = SECBUFFER_TOKEN, },
+        { .BufferType = SECBUFFER_EMPTY },
+        { .BufferType = SECBUFFER_EMPTY },
+    };
+    SecBufferDesc outbuf_desc = { SECBUFFER_VERSION, 2, outbuf };
+
+    SecBuffer inbuf[2] = { };
+    inbuf[0].cbBuffer = engine->inbound_len;
+    inbuf[0].pvBuffer = engine->inbound;
+    inbuf[0].BufferType = SECBUFFER_TOKEN;
+
+    SecBufferDesc inbuf_desc = { SECBUFFER_VERSION, 2, inbuf };
+
+    if (engine->status == SEC_I_CONTINUE_NEEDED ||
+        engine->status == SEC_E_INCOMPLETE_MESSAGE) {
+        ssize_t read = engine->read_fn(engine->io,
+                                      engine->inbound + engine->inbound_len,
+                                      sizeof(engine->inbound) - engine->inbound_len);
+        if (read > 0) {
+            engine->inbound_len += read;
+            UM_LOG(VERB, "read %zd bytes of handshake data", read);
+            inbuf[0].cbBuffer = (unsigned long)engine->inbound_len;
+            inbuf[0].pvBuffer = engine->inbound;
+        } else {
+            UM_LOG(ERR, "failed to read handshake data: %zd", read);
+            engine->handshake_st = TLS_HS_ERROR;
+            return engine->handshake_st;
+        }
+    }
+
+    char in[32000];
```

Comment:
would `32000` be better off as a #define somewhere? Why "almost 32k" not `2 << 14` or `32768`. I take it this is just 'big enough'? magic numbers always make me ask 'why'

Reply:
it's not needed at all, just leftover from some debugging